pipx/pipxrc.py

```diff
@@ -0,0 +1,196 @@
+        return self._val_or_default(self.pipxrc_info.package_or_url, default)
+
+    def get_install_pip_args(self, default: List[str]) -> List[str]:
+        return self._val_or_default(self.pipxrc_info.install.pip_args, default)
+
+    def get_install_venv_args(self, default: List[str]) -> List[str]:
+        return self._val_or_default(self.pipxrc_info.install.venv_args, default)
+
+    def get_install_include_dependencies(self, default: bool) -> bool:
+        return self._val_or_default(
+            self.pipxrc_info.install.include_dependencies, default
+        )
+
+    def get_venv_metadata(self, default: PipxVenvMetadata) -> PipxVenvMetadata:
+        return self._val_or_default(self.pipxrc_info.venv_metadata, default)
+
+    def get_injected_packages(
+        self, default: List[Dict[str, Any]]
+    ) -> List[Dict[str, Any]]:
+        if self.pipxrc_info.injected_packages is not None:
+            injected_packages = []
+            for package in self.pipxrc_info.injected_packages:
+                package_info = {"package": package}
+                package_info.update(
+                    self.pipxrc_info.injected_packages[package]._asdict()
+                )
+                injected_packages.append(package_info)
+            return injected_packages
+        else:
+            return default
+
+    def set_package_or_url(self, package_or_url: str) -> None:
+        # TODO 20190923: if package_or_url is a local path, we need to make it
+        #   an absolute path
+        self.pipxrc_info.package_or_url = package_or_url
+
+    def set_venv_metadata(self, venv_metadata: PipxVenvMetadata) -> None:
+        self.pipxrc_info.venv_metadata = venv_metadata
+
+    def set_install_options(
+        self, pip_args: List[str], venv_args: List[str], include_dependencies: bool
+    ) -> None:
+        self.pipxrc_info.install = InstallOptions(
+            pip_args=pip_args,
+            venv_args=venv_args,
+            include_dependencies=include_dependencies,
+        )
+
+    def add_injected_package(
+        self,
+        package: str,
+        pip_args: List[str],
+        verbose: bool,
+        include_apps: bool,
+        include_dependencies: bool,
+        force: bool,
+    ) -> None:
+        if self.pipxrc_info.injected_packages is None:
+            self.pipxrc_info.injected_packages = {}
+
+        self.pipxrc_info.injected_packages[package] = InjectedPackage(
+            pip_args=pip_args,
+            verbose=verbose,
+            include_apps=include_apps,
+            include_dependencies=include_dependencies,
+            force=force,
+        )
+
+    def write(self) -> None:
+        # If writing out, make sure injected_packages is not None, so next
+        #   successful read of pipxrc does not use default in
+        #   get_injected_packages()
+        if self.pipxrc_info.injected_packages is None:
+            self.pipxrc_info.injected_packages = {}
+
+        # TODO 20190919: raise exception on failure?
```

Comment:
Recommend logging a warning and mentioning the functionality that will not be expected to work, not raising an exception

Reply:
I was wondering about this one, because being unable to write into a pipx venv directory by pipx seemed like an internal error to me.  I couldn't think of a normal user-caused way this would happen.

I also think once users become accustomed to pipx remembering everything, the functionality being removed will feel more broken than it does now, which also motivated making this a more fatal error.

Reply:
By raise exception, you mean you are leaving a try/catch out of this and letting a bare exception be raised?

Reply:
And yeah I agree it would be weird if this directory couldn't be written to by the user. But anything is always possible when enough users are using it :).

It is possible the permissions on the directory were changed, or a different user created it than the user that's trying to modify it.

Reply:
Yes that was one thought.  Because if this were to ever happen it seemed like an internal error.  I didn't know how we could make a venv and write into it without also allowing a write of pipxrc.json into it also.

But just out of paranoia's sake I felt I should at least think about catching this because that's typically what you do with a written file.

I'm pretty flexible with this one (which is why the TODO) so I am happy to do what you suggest.  Just wanted to show my reasoning.

Reply:
I would go with the try/catch and log a warning describing what went wrong, what the impact to the user is, and what they can do to fix the situation.

Reply:
Sorry, try/except. I do a lot of JavaScript programming too :smile:

Reply:
Done.  With a `logging.warning`